src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp

```diff
@@ -59,70 +60,86 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
       ++distanceCalculations;
 
       // Update bounds, if this intra-cluster distance is smaller.
-      if (dist < minClusterDistances(i))
-        minClusterDistances(i) = dist;
-      if (dist < minClusterDistances(j))
-        minClusterDistances(j) = dist;
+      minClusterDistances(i) = std::min(minClusterDistances(i), dist);
+      minClusterDistances(j) = std::min(minClusterDistances(j), dist);
     }
   }
 
-  for (size_t i = 0; i < dataset.n_cols; ++i)
+  #pragma omp parallel
```

Comment:
I was thinking to myself, "wouldn't it be nice if we could just use `#pragma omp parallel for reduction(+: centroids)` or something like this?"  That would be much nicer code-wise than making the `threadNewCentroids` object and `threadCounts` objects and then updating them.  Then I read about OpenMP custom reductions:

https://stackoverflow.com/questions/59034114/custom-omp-reduction-on-stdmaps

Do you want to give that a try and see if it works?  Don't feel obligated, we could also do it in another PR or another time.  But it could be nice for simplifying not just this code but other code that uses similar patterns for the other k-means strategies.  I guess that we could define Armadillo-specific custom reductions in some core header file, and then just use it where relevant.